### PR TITLE
Port remote configuration option to nginx.

### PR DESCRIPTION
**Pull request by crowell**
Deleting other one that I broke squashing.

### CODE REVIEW

**Comment by jeffkaufman** [comment]
LGTM, like with https://github.com/pagespeed/ngx_pagespeed/pull/905

**Comment by oschaaf** [comment]
Looks like this broke something in starting up nginx with the native fetcher on:
```
#0  net_instaweb::NgxEventConnection::WriteEvent (this=0x0, type=type@entry=88 'X', sender=sender@entry=0x1543fd0)
    at /home/oschaaf/code/google/ngx_pagespeed/src/ngx_event_connection.cc:142
#1  0x000000000047f915 in net_instaweb::NgxEventConnection::WriteEvent (this=<optimized out>, sender=sender@entry=0x1543fd0)
    at /home/oschaaf/code/google/ngx_pagespeed/src/ngx_event_connection.cc:128
#2  0x0000000000492023 in net_instaweb::NgxUrlAsyncFetcher::Fetch (this=0x1543fd0, url=..., message_handler=0x14396d0, async_fetch=<optimized out>)
    at /home/oschaaf/code/google/ngx_pagespeed/src/ngx_url_async_fetcher.cc:218
#3  0x00000000006226c0 in net_instaweb::RateController::Fetch (this=0x1544200, fetcher=0x1543fd0, url=..., message_handler=0x14396d0, fetch=0x2934640)
    at net/instaweb/http/rate_controller.cc:298
#4  0x000000000062696d in net_instaweb::RateControllingUrlAsyncFetcher::Fetch (this=0x15441d0, url=..., message_handler=0x14396d0, fetch=0x2934640)
    at net/instaweb/http/rate_controlling_url_async_fetcher.cc:49
#5  0x00000000006280e9 in net_instaweb::UrlAsyncFetcherStats::Fetch (this=0x292e470, url=..., message_handler=0x14396d0, fetch=0x29345e0)
    at net/instaweb/http/url_async_fetcher_stats.cc:128
#6  0x00000000006189ac in net_instaweb::(anonymous namespace)::CacheFindCallback::Done (this=0x2933d00, find_result=net_instaweb::HTTPCache::kNotFound)
    at net/instaweb/http/cache_url_async_fetcher.cc:380
#7  0x000000000061cd9c in net_instaweb::HTTPCacheCallback::Done (this=0x2934050, backend_state=net_instaweb::CacheInterface::kNotFound)
    at net/instaweb/http/http_cache.cc:264
#8  0x000000000066a79f in net_instaweb::CacheInterface::Callback::DelegatedDone (this=0x2934050, state=net_instaweb::CacheInterface::kNotFound)
    at ./pagespeed/kernel/base/cache_interface.h:55
#9  0x000000000066aadd in net_instaweb::DelegatingCacheCallback::Done (this=0x29340d0, state=net_instaweb::CacheInterface::kNotFound)
    at pagespeed/kernel/cache/delegating_cache_callback.cc:53
#10 0x0000000000669e4f in net_instaweb::CacheStats::StatsCallback::Done (this=0x29340d0, state=net_instaweb::CacheInterface::kNotFound)
    at pagespeed/kernel/cache/cache_stats.cc:128
#11 0x0000000000633283 in net_instaweb::CacheInterface::ValidateAndReportResult (this=0x2754ef0, key=..., state=net_instaweb::CacheInterface::kNotFound, 
    callback=0x29340d0) at pagespeed/kernel/base/cache_interface.cc:58
#12 0x000000000066bb53 in net_instaweb::FileCache::Get (this=0x2754ef0, key=..., callback=0x29340d0) at pagespeed/kernel/cache/file_cache.cc:137
#13 0x00000000006697d1 in net_instaweb::CacheStats::Get (this=0x2755210, key=..., callback=0x2934050) at pagespeed/kernel/cache/cache_stats.cc:144
#14 0x000000000061a582 in net_instaweb::HTTPCache::Find (this=0x292e630, key=..., fragment=..., handler=0x14396d0, callback=0x2933d00)
    at net/instaweb/http/http_cache.cc:285
#15 0x00000000006197a1 in net_instaweb::CacheUrlAsyncFetcher::Fetch (this=0x7fffffffd660, url=..., handler=0x14396d0, base_fetch=0x2930f90)
    at net/instaweb/http/cache_url_async_fetcher.cc:584
#16 0x00000000005860ad in net_instaweb::ServerContext::FetchRemoteConfig (this=0x15a00d0, url=..., timeout_ms=1500, implicit_cache_ttl_ms=300000, on_startup=true, 
    request_ctx=...) at net/instaweb/rewriter/server_context.cc:1341
#17 0x00000000005838cd in net_instaweb::ServerContext::GetRemoteOptions (this=0x15a00d0, remote_options=0x292f0f0, on_startup=true)
    at net/instaweb/rewriter/server_context.cc:925
#18 0x000000000050e3a3 in net_instaweb::RewriteDriverFactory::InitServerContext (this=0x14307d0, server_context=0x15a00d0)
    at net/instaweb/rewriter/rewrite_driver_factory.cc:573
#19 0x0000000000613966 in net_instaweb::SystemServerContext::ChildInit (this=0x15a00d0, factory=0x14307d0) at pagespeed/system/system_server_context.cc:286
#20 0x00000000006074da in net_instaweb::SystemRewriteDriverFactory::ChildInit (this=0x14307d0) at pagespeed/system/system_rewrite_driver_factory.cc:288
#21 0x0000000000483d32 in net_instaweb::(anonymous namespace)::ps_init_child_process (cycle=0x141aac0)
    at /home/oschaaf/code/google/ngx_pagespeed/src/ngx_pagespeed.cc:3047
#22 0x000000000042eddb in ngx_single_process_cycle (cycle=cycle@entry=0x141aac0) at src/os/unix/ngx_process_cycle.c:309
#23 0x000000000040d541 in main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:404
```

File: test/nginx_system_test.sh [hunk 1]
```diff
@@ -1263,6 +1277,10 @@ OUT=$(cat "$ERROR_LOG" \
     | grep -v "\\[error\\].*Failed to make directory*" \
     | grep -v "\\[error\\].*Could not create directories*" \
     | grep -v "\\[error\\].*opening temp file: No such file or directory.*" \
+    | grep -v "\\[error\\].*unexpected response.*" \
```
Comment (oschaaf) [inline test/nginx_system_test.sh]: @crowell was (re-)adding this line really needed? I was hoping that https://github.com/pagespeed/ngx_pagespeed/pull/922 would have fixed this -- have you been testing with that change included?

Comment (crowell) [inline test/nginx_system_test.sh]: re adding those lines is not needed. I can remove them.  I didn't see that
error on startup with the native fetcher on my tests before check in. I'll
take a look today.

**Comment by crowell** [comment]
@oschaaf pushed a new pr to remove that line, tests still pass.

**Comment by oschaaf** [comment]
@crowell just upgraded to the latest on trunk to test some upcoming things, and noted that under valgrind, one of these tests will flake

**Comment by oschaaf** [comment]
@crowell No, wait, without valgrind it seems to fail as well:
```
Sleeping so that cache will expire
      Fetching http://remote-config-failed-fetch.example.com/mod_pagespeed_test/forbidden.html    -o /tmp/mod_pagespeed_test.oschaaf/nginx/fetched_directory/wget_output.txt                                                   -O /tmp/mod_pagespeed_test.oschaaf/nginx/fetched_directory/wget/fetch_until_output.28980 until $(fgrep -c <!--) = 0
wget -q    -o /tmp/mod_pagespeed_test.oschaaf/nginx/fetched_directory/wget_output.txt                                                   -O /tmp/mod_pagespeed_test.oschaaf/nginx/fetched_directory/wget/fetch_until_output.28980 http://remote-config-failed-fetch.example.com/mod_pagespeed_test/forbidden.html and checking with fgrep -c <!--
........
/home/oschaaf/code/google/mod_pagespeed//src/pagespeed/system/system_test.sh: line 739: kill: (31410) - No such process
FAILed Input: Line:739
/home/oschaaf/code/google/mod_pagespeed/src/pagespeed/system/system_test.sh:739: source
/home/oschaaf/code/google/ngx_pagespeed/test/nginx_system_test.sh:339: main
     failure at line 339
in 'second remote config fetch fails, cached value still applies.'
FAIL.
With serf fetcher setup.

```
I'll see if I can dig up some more info, it might be something I broke in the test script

**Comment by crowell** [comment]
I'll take a look on Monday, the tests seem to be fine from the times I've run them though

**Comment by oschaaf** [comment]
@crowell I'll have a look too, I'm not sure yet what was causing this. I commented this test, and all the others are passing. I've seen this test run fine earlier.

**Comment by oschaaf** [comment]
@crowell After killing a stray 'nc' process that was orphaned by the test script, the failure stopped happening. Perhaps I killed the test script at an unfortunate moment some time earlier.

**Comment by crowell** [comment]
ah, that sounds like it could be the case.  it should be possible to do
something nicer, like killing things listening on the used ports to begin
with though.

I'll look into adding that tomorrow!

On Sun, Mar 15, 2015 at 4:55 PM, Otto van der Schaaf <
notifications@github.com> wrote:

> @crowell <https://github.com/crowell> After killing a stray 'nc' process
> that was orphaned by the test script, the failure stopped happening.
> Perhaps I killed the test script at an unfortunate moment some time earlier.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pagespeed/ngx_pagespeed/pull/925#issuecomment-81235917>
> .
>



-- 
Jeff Crowell
https://github.com/crowell

**Comment by crowell** [comment]
@oschaaf I have a fix for this waiting on review in the modpagespeed tests.

**Pull request merged**